File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set<ListNode*> node;
        while(head!=NULL){
            if(node.find(head)==node.end())
                node.insert(head);
            else
                return head;
            head=head->next;
        }
        return NULL;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        HashSet<ListNode> s=new HashSet<>();
        if(head==null){
            return null;
        }
        while(head.next!=null){
            if(s.contains(head)){
                return head;
            }
            s.add(head);
            head=head.next;
        }
        return null;
    }
}